Source/WTF/wtf/ThreadSafeWeakHashSet.h

```diff
@@ -92,55 +92,64 @@ class ThreadSafeWeakHashSet final {
         Locker locker { m_lock };
         ControlBlockRefPtr retainedControlBlock { &value.controlBlock() };
-        if (!retainedControlBlock)
-            return;
+        ASSERT(retainedControlBlock);
         amortizedCleanupIfNeeded();
-        m_map.add(static_cast<const T*>(&value), WTFMove(retainedControlBlock));
+        m_set.add(std::make_pair(WTFMove(retainedControlBlock), &value));
     }
 
-    template<typename U, std::enable_if_t<std::is_convertible_v<U*, T*>>* = nullptr>
-    bool remove(const U& value)
+    template<typename U>
+    bool remove(const U& value) requires (std::is_convertible_v<U*, T*>)
     {
         Locker locker { m_lock };
         amortizedCleanupIfNeeded();
-        auto it = m_map.find(static_cast<const T*>(&value));
-        if (it == m_map.end())
+        // FIXME: If there's no weak refs then it can't be in our table. In that case
+        // there's no point in potentially allocating a ControlBlock.
+        auto it = m_set.find(std::make_pair(&value.controlBlock(), &value));
+        if (it == m_set.end() || !it->first)
             return false;
-        bool wasDeleted = it->value && it->value->objectHasStartedDeletion();
-        bool result = m_map.remove(it);
-        return !wasDeleted && result;
+        bool wasDeleted = it->first->objectHasStartedDeletion();
+        bool result = m_set.remove(it);
+        ASSERT_UNUSED(result, result);
+        return !wasDeleted;
     }
 
     void clear()
     {
         Locker locker { m_lock };
-        m_map.clear();
+        m_set.clear();
         cleanupHappened();
     }
 
-    template<typename U, std::enable_if_t<std::is_convertible_v<U*, T*>>* = nullptr>
-    bool contains(const U& value) const
+    template<typename U>
+    bool contains(const U& value) const requires (std::is_convertible_v<U*, T*>)
     {
         Locker locker { m_lock };
         amortizedCleanupIfNeeded();
-        auto it = m_map.find(static_cast<const T*>(&value));
-        if (it == m_map.end())
+        // FIXME: If there's no weak refs then it can't be in our table. In that case
+        // there's no point in potentially allocating a ControlBlock.
+        auto it = m_set.find(std::make_pair(&value.controlBlock(), &value));
+        if (it == m_set.end() || !it->first)
             return false;
-        return it->value && !it->value->objectHasStartedDeletion();
+
+        bool wasDeleted = it->first->objectHasStartedDeletion();
+        if (wasDeleted)
+            m_set.remove(it);
+        return !wasDeleted;
     }
 
     bool isEmptyIgnoringNullReferences() const
     {
         Locker locker { m_lock };
         amortizedCleanupIfNeeded();
-        for (auto& controlBlock : m_map.values()) {
-            if (!controlBlock->objectHasStartedDeletion())
+        // FIXME: This seems like it should remove any stale entries it finds.
+        for (auto& pair : m_set) {
+            if (!pair.first->objectHasStartedDeletion())
```

Comment:
Does this also need to check if `pair.first` as we do in the `remove` and `contains` implementations above?

Reply:
Actually I think none of the `!.*first` checks are needed since we don't put null ControlBlocks into the set.